numpyro/hmc_util.py

```diff
@@ -775,9 +775,30 @@ def single_chain_init(key, only_params=False):
                     potential_energy(seeded_model, model_args, model_kwargs, inv_transforms),
                     jax.partial(transform_fn, inv_transforms))
 
+    _, potential_fn, constrain_fn = single_chain_init(rng if rng.ndim == 1 else rng[0])
+
+    def single_chain_init_validated(key):
```

Comment:
Is it possible to make this a separate external function `find_valid_initial_params(key, potential_fn, strategy='uniform', prior_params=None)` which can be called from within `single_chain_init` by default? That way, we can also repurpose it for autoguides.

Reply:
For `uniform` strategy, it is quite straightforward. But for `prior` strategy, we have to get traces from the model. I think that I can revise the logic here a bit to match your suggestion.

Reply:
@neerajprad I have implemented `find_valid_initial_params` in `infer_util` together with various init_strategy we have so far: `feasible`, `uniform`, `prior`, `median`. Looking like it works as expected (modulo issue #241). Does this align with your suggestion?

Reply:
This is looking great!